### DISCUSSION TITLE
Chat QA Slack通知機能

### DISCUSSION THREAD

**Discussion by jacopen** [cloudnativedaysjp/dreamkast#1578] [Ideas]
https://github.com/cloudnativedaysjp/cndt2022/discussions/183  でディスカッションされているものを自動化する機能

## Motivation
現在のDreamkast UI Chatだと、参加者が質問を投げても登壇者が気づく前に次のセッションに行ってしまったり、日付が変わって閲覧できなくなったりする。

登壇者が質問に対してセッション中、もしくはセッション後に返答ができるように、CNDT Slackで登壇者にQAをメンションする機能が欲しい

**Comment by jacopen** [comment]
Chat messageでQAだったら何らしかのアクションをフックするということはそれほど難しくない。
https://github.com/cloudnativedaysjp/dreamkast/blob/main/app/controllers/api/v1/chat_messages_controller.rb#L19-L39
のメソッド内でmessage_typeが1(QA) だったら何かをする　っていう分岐を入れるだけで良い。

なのでSlackのincoming webhookでメッセージを送るような処理を追加するのが一番単純な実装。

ただし、こういう同期的な処理はもし仮にSlack側のAPIのレスポンスが遅かったりすると処理が詰まってしまう可能性がある。
なので、workerを使って非同期にメッセージを送るようにしたほうが安牌

**Reply by jacopen** [reply]
Slackに送るメッセージ案

```
@登壇者 Track A - XXXについて語るセッション  に質問がありました
Q. XXXについて、YYYはどういう仕組みになっているんでしょう？
```

**Reply by jacopen** [reply]
現在のDkのDBから、
- Track名
- Talk名
- 質問内容
は簡単に抽出できる。 登壇者と登壇者のSlack名を結びつけるデータが現状は無いため、ここについては仕組みの追加が必要

**Comment by hrk091** [comment]
何も背景がわからない無邪気な発言で申し訳ないのですが、chat機能を自前で具備せず、単純にtwitterのハッシュタグ検索したタイムラインを表示するのはだめなのでしょうか？（なんとなく、過去に検討していてNGな理由があったのかなと思っております）

「chat -> slack -> twitterに書くように誘導、というのを頑張る」よりも、「やり取りはすべてtwitterで完結させて、dkはそのハッシュタグのものを単純に表示する」としたほうが、実装コストも下がり利便性もあがり、得られるメリットが大きいように感じました。うろ覚えですが、JANOGも同じような方式を取っていた気がします。

ハッシュタグ検索のタイムライン表示は少々お金かインテグレーションコストがかかりそうですが、動線整理・運用にコストがかかっていることも鑑みて、検討しても良いかもしれません。

**Reply by jacopen** [reply]
そうですねー、歴史的な話、参加者の反応、技術的な話いろいろあって現在の形になっています。
まず考え方としてTwitterにまとめるのも間違いではないと思っています。

- イベントに参加していない人たちにも盛り上がり感を伝えられる
- Twitterと内部チャットどっちにかけばいいの？という混乱を防げる
- 慣れたツールで書き込みが出来る

というメリットがあると思っていて、特に1つめが大きいですね。

一方で考えなければいけない点もあります。

- QAをどうするか

現在の機能になったのはこれが大きいです。一般的な勉強会でも、Twitterで自由につぶやきつつも質問がある場合はSlidoにみたいなパターンが多いと思うんですが、Dkも初版ではSlido埋め込みを使っていました。ただ、これがめちゃくちゃUXが悪くて、評判が悪かったのです。ということで、質問と返答ができる仕組みを内部に持とうということで実装しました。
その後、『Q&Aの機能とチャットを一体化させたら便利なんじゃね？』 というアイディアがあり、チャットも兼ねる形になったのです。

もうひとつ

- Twitterをやっていない、もしくは秘密にしたい人も案外多い

という点も考慮したいポイントです。開催後アンケートを見ると、「アカウントを晒すことなく匿名でチャットできるのがよかった」という反応もあったりして、意外と需要あるんだなと。なので、両方の選択肢を提供しているという状態です。

なので今のところはチャット続投で考えているんですが、ハッシュタグ検索のタイムライン表示はそれはそれでUIに実装したい機能です。で、それはTwitter APIが必要で、そのあたりの申請は過去に行っていたりします。
費用的にはなんとかなるのですが、実はかつてcloudnativedaysのTwitterアカウントは凍結を食らったことがありまして。。。
で、そも凍結の影響か、API利用に制限がかかっていたりするのです。なので、Twitterタイムラインの表示やツイートの自動化にはアカウントを別に作り直す必要がありそうで、それがめちゃくちゃしんどい状況だったりします･･･

**Reply by hrk091** [reply]
詳細な回答ありがとうございます！過去背景すごく理解できました。
うーん、なるほどです。悩ましいですね。。。

そもそも、問題を整理すると、
- 登壇者が、質問が来ていることに気づけない
- 質問者が、回答が来ていることに気づけない
- イベント終了後にチャットを見ることができない
- （chatの場所がわかりにくい？）

あたりなので、twitterやslackに頼らずに独自で匿名で質問できることにニーズがあるならば、dk上でうまく解けるとよさそうですね。
twitterやslackを経由せずに、質問が来ていることがdk上（かつ分かりやすく）認識することができれば良いので、何か方法はありそう。

https://github.com/cloudnativedaysjp/cndt2022/discussions/183#discussioncomment-3943640

でつかまんさんが書いてくれてる案か、あるいはnotify機能を設けるとかかなあ

**Reply by jacopen** [reply]
あれ、言われて気づいたんですが今はスピーカーダッシュボードもあるし、セッションのQuestionを対象スピーカーのスピーカーダッシュボードに表示するだけなら瞬殺で実装できるな？？？なんで思いつかなかったのか、、、

**Comment by takaishi** [comment]
セッション終了後もQAについて一覧が見えたり、質問者とやりとりができるような機能が必要ですねえ。やりたいと思いつつまだできていない。
